File: renderer/src/ast.ts
/**
 * PARSEC AST Type Definitions
 * Mirror of Rust AST structures
 */

export interface Program {
  app?: AppNode;
  imports: Import[];
  components: ComponentDef[];
}

export interface Import {
  path: string;
}

export interface ComponentDef {
  name: string;
  props: string[];
  children: Node[];
}

export interface AppNode {
  meta?: MetaNode;
  views: ViewNode[];
}

export interface MetaNode {
  attrs: Record<string, AttributeValue>;
}

export interface ViewNode {
  path?: string;
  children: Node[];
}

export type Node =
  | TextNode
  | BindingNode
  | ElementNode
  | ComponentNode
  | ConditionalNode
  | LoopNode;

export interface TextNode {
  type: 'text';
  content: string;
}

export interface BindingNode {
  type: 'binding';
  expression: string;
}

export interface ElementNode {
  type: 'element';
  tag: string;
  attrs: Record<string, AttributeValue>;
  children: Node[];
}

export interface ComponentNode {
  type: 'component';
  name: string;
  props: Record<string, AttributeValue>;
  children: Node[];
}

export interface ConditionalNode {
  type: 'conditional';
  condition: Binding;
  thenBody: Node[];
  elseBody?: Node[];
}

export interface LoopNode {
  type: 'loop';
  iterable: Binding;
  itemName: string;
  children: Node[];
}

export interface Binding {
  expression: string;
}

export type AttributeValue =
  | string
  | number
  | boolean
  | Binding
  | AttributeValue[]
  | Record<string, any>;

export interface ParseError {
  type: 'error';
  message: string;
  position: number;
}

/**
 * Type guards
 */
export function isTextNode(node: Node): node is TextNode {
  return 'type' in node && node.type === 'text';
}

export function isElementNode(node: Node): node is ElementNode {
  return 'type' in node && node.type === 'element';
}

export function isComponentNode(node: Node): node is ComponentNode {
  return 'type' in node && node.type === 'component';
}

export function isConditionalNode(node: Node): node is ConditionalNode {
  return 'type' in node && node.type === 'conditional';
}

export function isLoopNode(node: Node): node is LoopNode {
  return 'type' in node && node.type === 'loop';
}

export function isBindingNode(node: Node): node is BindingNode {
  return 'type' in node && node.type === 'binding';
}

export function isBinding(value: AttributeValue): value is Binding {
  return typeof value === 'object' && 'expression' in value && !Array.isArray(value);
}

File: parser/src/ast.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Program {
    pub app: Option<AppNode>,
    pub imports: Vec<Import>,
    pub components: Vec<ComponentDef>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Import {
    pub path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ComponentDef {
    pub name: String,
    pub props: Vec<String>,
    pub children: Vec<Node>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AppNode {
    pub meta: Option<MetaNode>,
    pub views: Vec<ViewNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MetaNode {
    pub attrs: HashMap<String, AttributeValue>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ViewNode {
    pub path: Option<String>,
    pub children: Vec<Node>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "content")]
pub enum Node {
    Text(String),
    Binding(Binding),
    Element {
        tag: String,
        attrs: HashMap<String, AttributeValue>,
        children: Vec<Node>,
    },
    Component {
        name: String,
        props: HashMap<String, AttributeValue>,
        children: Vec<Node>,
    },
    Conditional {
        condition: Binding,
        then_body: Vec<Node>,
        else_body: Option<Vec<Node>>,
    },
    Loop {
        iterable: Binding,
        item_name: String,
        children: Vec<Node>,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Binding {
    pub expression: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum AttributeValue {
    String(String),
    Number(f64),
    Boolean(bool),
    Binding(Binding),
    List(Vec<AttributeValue>),
    Object(HashMap<String, AttributeValue>),
}

impl AttributeValue {
    pub fn to_string(&self) -> String {
        match self {
            AttributeValue::String(s) => s.clone(),
            AttributeValue::Number(n) => n.to_string(),
            AttributeValue::Boolean(b) => b.to_string(),
            AttributeValue::Binding(b) => format!("{{{}}}", b.expression),
            AttributeValue::List(l) => format!("[{:?}]", l),
            AttributeValue::Object(_) => "{...}".to_string(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ast_creation() {
        let program = Program {
            app: None,
            imports: vec![],
            components: vec![],
        };
        assert!(program.app.is_none());
    }
}
